core/arch/arm/plat-rcar/romapi.c

```diff
@@ -0,0 +1,140 @@
+			break;
+		default:	/* Newer H3 versions use unified table */
+			index = 3;
+			break;
+		}
+		break;
+	case PRR_PRODUCT_M3W:
+		switch (rcar_prr_value & PRR_CUT_MASK) {
+		case PRR_CUT_10:	/* M3 ES1.0 */
+			index = 2;
+			break;
+		default:	/* Newer M3 versions use unified table */
+			index = 3;
+			break;
+		}
+		break;
+	default:			/* All other SoCs use unified table */
+		index = 3;
+		break;
+	}
+
+	return index;
+}
+
+/* implemented in romapi_call.S */
+extern uint32_t __plat_call_romapi(paddr_t sp, paddr_t func, uint64_t arg1,
+				   uint64_t arg2, uint64_t arg3);
+
+static uint32_t plat_call_romapi(paddr_t func, uint64_t arg1,
+				 uint64_t arg2, uint64_t arg3)
+{
+	/*
+	 * If MMU is enabled, we need to use trampoline function that will
+	 * disable MMU and switch stack pointer to physical address. On other
+	 * hand, if MMU is disabled, we can call the ROM function directly.
+	 */
+	if (cpu_mmu_enabled()) {
+		uint32_t (*fptr)(paddr_t func, uint64_t arg1, uint64_t arg2,
+				 uint64_t arg3);
+
+		/*
+		 * __plat_call_romapi() is identity mapped and we should call it
+		 * at identity address, otherwise it will experience problem
+		 * when MMU will be disabled.
+		 */
+		fptr = (void *)virt_to_phys(__plat_call_romapi);
+
+		return fptr(func, arg1, arg2, arg3);
+	} else {
+		uint32_t (*fptr)(uint64_t arg1, uint64_t arg2, uint64_t arg3);
+
+		fptr = (typeof(fptr))func;
+
+		return fptr(arg1, arg2, arg3);
+	}
+}
+
+static paddr_t va2pa(void *ptr)
+{
+	if (cpu_mmu_enabled())
+		return virt_to_phys(ptr);
+	else
+		return (paddr_t)ptr;
+}
+
+static const paddr_t romapi_getrndvector[] = {
+	0xEB10DFC4,	/* H3 1.0/1.1, needs confirmation */
+	0xEB117134,	/* H3 2.0 */
+	0xEB11055C,	/* M3 1.0/1.05, needs confirmation */
+	0xEB100188,	/* H3 3.0, M3 1.1+, M3N, E3, D3, V3M 2.0 */
+};
+
+uint32_t plat_rom_getrndvector(uint8_t rndbuff[PLAT_RND_VECTOR_SZ],
+			       uint8_t *scratch, uint32_t scratch_sz)
+{
+	uint32_t ret = -1;
+	paddr_t func_addr = romapi_getrndvector[get_api_table_index()];
+	paddr_t rndbuff_pa = va2pa(rndbuff);
+	paddr_t scratch_pa = va2pa(scratch);
+
+	assert(scratch_sz >= 4096);
+	assert(rndbuff_pa % 8 == 0);
+	assert(scratch_pa % 8 == 0);
+
+	ret = plat_call_romapi(func_addr, rndbuff_pa, scratch_pa, scratch_sz);
+
+	/*
+	 * ROM code is called with MMU turned off, so any accesses to rndbuff
+	 * are not affected by data cache. This can lead to two problems:
+	 *
+	 * 1. Any prior writes can be cached but may not reach memory. So staled
+	 * values can be flushed to memory later and overwrite new data written
+	 * by ROM code. This includes stack as well.
+	 *
+	 * 2. ROM code will write new data to the buffer, but we may see old,
+	 * cached values.
+	 *
+	 * ROM code wrapper will issue dcache_op_all(DCACHE_OP_CLEAN). This will
+	 * ensure that all writes reached memory. After the call we need to
+	 * invalidate the cache to see new data.
+	 *
+	 * We are not accessing scratch area, so no need to do cache maintenance
```

Comment:
Then I believe the scratch buffer should be cache line aligned.

Reply:
Hm, not sure I get this.

This buffer never is being read or written by OP-TEE. ROM code needs it to store some own temporary data. So, why it should be cache line aligned?

Reply:
What if there's a partial overlap with a cache line and some cached memory? If so you may have sudden corruptions in case that cache line is dirty and written into memory.

Reply:
Ah, I see. I had the same exact problem with a stack. So, `__plat_call_romapi` clears all data cache before jumping to ROM code. I do understand that this is sub-optimal, but I didn't found better solution.

Reply:
If you make the memory areas cache line aligned too it should be water tight at least.
That stuff works is always a priority. ;-)